hybrid/compartment_examples/inter_comp_call/secure/main.c

```diff
@@ -63,6 +63,7 @@ extern void comp_f_fn();
 extern void comp_g_fn();
 extern void *comp_f_fn_end;
 extern void *comp_g_fn_end;
+extern void *switch_compartment_end;
```

Comment:
The types of these `_end` variables aren't really correct. These are the types that they point to, which here is just a zero-sized part of the code (`.text` section). There isn't a good, correct choice, but perhaps representing them as a function pointer is best for this example. Just don't call them; that won't end well. (It shouldn't be possible if the compartment is set up correctly.)

I think this caused your earlier question about taking the address. It's surprising to take the address of a global `void*` but not so much to take the address of a function, especially given the arithmetic where this is used.

Reply:
I was thinking address values are usually represented as `void*`, but if you think functions are better, I have nothing against that. Amended in https://github.com/capablevms/cheri-examples/pull/56/commits/1b46b9cebdf195bcedad7e24ee673a9450701388.

Reply:
They aren't addresses, though. If anything, they're `void`s. The label points to the value (which has zero size) and you take the address (`&switch_compartment_end`) to do some pointer arithmetic.

Reply:
I'll maybe have to read more on this, but I believe the current incarnation should be sufficient at the time, until we find a "proper" way (if one such way exists). Or maybe improve how we calculate function sizes in #54 .

Reply:
@jacobbramley Are you suggesting using something other than `void *` e.g. `char *` or `u8*`?

Reply:
I would declare it as a function ...
```
extern void switch_compartment_end();
```
... because a function has no size as far as C is concerned. All the other things do (except `void`, but I think that'll be rejected). This also works better in C64 because bit zero can be set in _both_ so that the size calculation will still work correctly. It doesn't matter now, but might do if this is adapted later.

This also makes a little more sense in the arithmetic semantics, since the `&` becomes implicit and then you're subtracting one function pointer from another, rather than a function pointer from a `void**`:

```
//                                 v-- '&' elided here because it's implicit for functions.
size_t switcher_size = `(uintptr_t) switch_compartment_end - (uintptr_t) switch_compartment`;
```

All of these options are _undefined behaviour_, so it's just about ergonomics. I was simply confused by the mix of types. That said, it's not really that important, and we've probably spent too long discussing it already :-)

Reply:
Ah, I think I see what you mean. @0152la maybe worth making this change, as I don't think it's a big one?

Reply:
This should have been done in https://github.com/capablevms/cheri-examples/commit/1b46b9cebdf195bcedad7e24ee673a9450701388, as per the above, unless I'm misunderstanding?

Reply:
I think `1b46b9c` made them function pointers. I.e. so C thinks there's an 8-byte function pointer at `&switch_compartment_end`, which there isn't.

Reply:
I think @jacobbramley is suggesting `s/void */void/`.

Reply:
I see; I was confused by the small detail. I think what we want is in https://github.com/capablevms/cheri-examples/pull/56/commits/420bddc50eba52cfd7d15f31e917ccdc38486d04.

Reply:
I think you also need to `s/extern void *comp_f_fn_end();/extern void comp_f_fn_end();/`?

Reply:
I think they are as well [1]?

https://github.com/capablevms/cheri-examples/pull/56/files#diff-6d24dc24a1e7111722091fada0a3b1105a2b592c8eb46c5d327a3fda9ebf61c8R64-R65

Reply:
Oops, I must have been looking at the wrong thing -- sorry!